src/mlpack/methods/random_forest/bootstrap.hpp

```diff
@@ -18,30 +18,234 @@ namespace mlpack {
+    const arma::uvec& indices,
+    const arma::vec&  concurrency)
+  {
+    arma::vec       avg(indices.n_rows, arma::fill::zeros);
+    const arma::vec invConcurrency(
+      arma::vec(concurrency.n_rows, arma::fill::ones) / concurrency);
+
+    // In production code this loop is only entered once because
+    // then indices.n_rows == 1.
+    for (arma::uword i(0); i < indices.n_rows; ++i) {
+      const arma::uword start(intervals(0, indices[i]));
+      const arma::uword end(intervals(1, indices[i]));
+
+      avg[i] = arma::accu(invConcurrency.rows(start, end)) /
+        (end - start + 1);
+    }
+
+    return avg;
+  }
+
+  /**
+   * Compute probabilities of the next draw for each observation.
+   *
+   * @param[in] phi List of previously drawn observations.
+   * @param[in] phiSize Number of elements in use in @p phi. To
+   *                    avoid reallocations, @p phi has been
+   *                    overallocated already.
+   * @param[in] concurrency Concurrency of the samples selected
+   *                        by @p phi so far.
+   * @param[in] intervals See ComputeAverageUniqueness().
+   * @return The probabilities for each observation to be drawn in the next
+   *         iteration.
+   */
+  static arma::vec ComputeNextDrawProbabilities(
+    arma::uvec&       phi,
+    arma::uword       phiSize,
+    arma::vec&        concurrency,
+    const IndMatType& intervals)
+  {
+    // phi may have less rows than intervals has columns as
+    // phi is grown in multiple steps.
+    assert(phi.n_rows <= intervals.n_cols && phiSize < phi.n_rows);
+
+    // Only in the test code we want to compute the full vector
+    // of average uniqueness. Here, we are always only interested
+    // in the average uniqueness of the latest value that is
+    // added to phi.
+    const arma::uvec indices(1u, arma::fill::value(phiSize));
+    arma::vec        avg(intervals.n_cols);
+
+    for (arma::uword i(0); i < avg.size(); ++i) {
+      // Temporarily assume another concurrency for the current i.
+      concurrency.subvec(intervals(0, i), intervals(1, i)) += 1.0;
+
+      phi[phiSize] = i;
+      avg[i] = ComputeAverageUniqueness(
+        intervals.cols(phi.rows(0, phiSize)),
+        indices,
+        concurrency).back();
+
+      concurrency.subvec(intervals(0, i), intervals(1, i)) -= 1.0;
+    }
+
+    return avg / arma::sum(avg);
+  }
+
+  /**
+   * Compute the samples of the next draw.
+   *
+   * @param[in] colCount Number of data points in the dataset and
+   *                     the number of samples to draw respectively.
+   *
+   * @return A list of indices referring to the observations that should
+   *         be sampled.
+   */
+  arma::uvec ComputeSamples(arma::uword colCount) const
+  {
+    // Randomly sample columns from the intervals matrix.
+    DiscreteDistribution d(intervals.n_cols);
+    arma::uvec           phi(colCount);
+    arma::vec            concurrency(intervals.max() + 1, arma::fill::zeros);
```

Comment:
Shouldn't we fill this with the actual concurrency values here?

Reply:
@rcurtin , Here, the concurrency is not known yet. As we are randomly adding elements to phi, the concurrency changes. This vector we only hold as an optimization to compute the concurrency in each step.